### PR TITLE
Support for the program data address space option of LLVM's Target Datalayout

### PR DESCRIPTION
**Pull request by TimNN**
This was introduced recently (specifically, for AVR, cc @dylanmckay).

(I came up with this when attempting to run [avr-rust](https://github.com/avr-rust/rust) rebased on the latest [rust-lang](https://github.com/rust-lang/rust) commits. If this requires a different design, some additional discussions, or is not something to pursue right now, I'd be happy to close this PR).

Note that this somewhat overlaps with @DiamondLovesYou's #51576, I think, although the implementation here is significantly simpler: Since the address space applies to _all_ program data, we can just check the pointee's type whenever we create an LLVM pointer type. If it is a function we use the program data address space; if not we use the default address space.

cc @eddyb, who has been reviewing #51576

Ref: https://llvm.org/docs/LangRef.html#data-layout

### CODE REVIEW

File: src/librustc_codegen_llvm/type_.rs
Comment (nikomatsakis) [inline src/librustc_codegen_llvm/type_.rs]: It is possible to "mis-use" this, I assume, for example via transmutes or something? Can we wind up with "valid Rust code" giving weird LLVM assertion failures or other errors?

**Comment by nikomatsakis** [comment]
cc @rust-lang/compiler — seems like an ad-hoc extension, nominating for brief T-compiler discussion to decide best path

File: src/librustc_codegen_llvm/type_.rs
Comment (TimNN) [reply src/librustc_codegen_llvm/type_.rs]: That's a good point. I'll see what happens if I transmute a few things to a function pointer.

**Comment by TimNN** [comment]
> seems like an ad-hoc extension

Yeah, this is definitely a bit ad-hoc, however overall the approach doesn't seem too bad to me. In any case, I came up with this while trying to get a target to work that isn't even officially supported yet. I'd be totally happy if you said this requires some more design work in the future.

**Comment by nikomatsakis** [comment]
@TimNN to be clear, I didn't mean "ad hoc" in the pejorative sense. Just that it's one of those "small features" that we often decide without an RFC -- it might however warrant a feature gate or a tracking issue or some such thing.

**Comment by DiamondLovesYou** [comment]
My patch can be easily extended to support this. But I think the space should be called the instruction address space, since "program memory" is kinda ambiguous (even though this is the term LLVM uses).

@eddyb wanted any value to be cast into the flat address space before that value is used (ie uses of a `global` would actually use an `addrspacecast`-ed const expr), in order to reduce the patch's complexity footprint. @TimNN would this be possible for the AVR target machine? The initial version I've whipped up doesn't force this backflip (it uses the correct space for functions when creating LLVM types from our types), but @eddyb might want them to go through a flat space cast round trip.

**Comment by TimNN** [comment]
>  think the space should be called the instruction address space

I'd be fine with that, as you noticed I was just copying LLVM terminology.

> @eddyb wanted any value to be cast into the flat address space before that value is used (ie uses of a `global` would actually use an `addrspacecast`-ed const expr), in order to reduce the patch's complexity footprint. @TimNN would this be possible for the AVR target machine?

I don't know. I understand how my patch works but overall I'm mostly unfamiliar with the involved concepts. IIRC, however, LLVM was unhappy casting a function to a function pointer in an incorrect addrspace (In case that information helps you).

---

In any case, if you think that the functionality added here could be better supported with your patch, I'd be happy to wait for that to be merged. (I'd probably open this PR against the `avr-rust` repo for others to use in the meantime, but for me it is in no way urgent to get this submitted).

File: src/librustc_codegen_llvm/type_.rs
Comment (eddyb) [inline src/librustc_codegen_llvm/type_.rs]: I really want to avoid us introspecting LLVM types. I think we should either not use non-flat address space pointers *or* we should special-case `fn` pointers - from a quick search, I found only 2 places that create them:
https://github.com/rust-lang/rust/blob/40123a1257f7645bee1c6d3b6eb8c2f6eebd206c/src/librustc_codegen_llvm/type_of.rs#L268
https://github.com/rust-lang/rust/blob/40123a1257f7645bee1c6d3b6eb8c2f6eebd206c/src/librustc_codegen_llvm/meth.rs#L42

Since these are both calling `.ptr_to()` on the result of `FnType::llvm_type`, a helper method on `FnType` can probably be added.

File: src/librustc_target/abi/mod.rs
Comment (eddyb) [inline src/librustc_target/abi/mod.rs]:
```suggestion
                [p] if p.starts_with("P") =>
```

**Comment by DiamondLovesYou** [comment]
@TimNN Well, I've push a change to support the feature needed in this PR, so feel free to give it a try.

> I don't know. I understand how my patch works but overall I'm mostly unfamiliar with the involved concepts. IIRC, however, LLVM was unhappy casting a function to a function pointer in an incorrect addrspace (In case that information helps you).

It does help. As I said, my patch put the vtable entries and all declared functions in the instruction address space, so that should work. I just realized it would not, however, do the same for function ptrs cast from non fn ptr values, like integers (perhaps a bit of an edge case). Currently, it tries to cast the value to the flat address space just like any other ptr.

But, baring that, it should work for AVR with a suitable target spec. If you'd like to give my patch a shot, let me know how it goes!

I would definitely prefer to have my patch merged instead; it (now) supports (that int -> fn ptr issue not withstanding) this address space, and can compile Rust which then runs on an AMDGPU.

**Comment by TimNN** [comment]
Thanks for your reviews, @eddyb & @nikomatsakis! 

I'm not sure what you think, but I feel like @DiamondLovesYou's PR #51576 is the way to go in the long term. 

I however don't know what the timeline would be to get that merged (or if you maybe have concerns that would prevent merging it). 

I see two ways forward:

1) close this for now until a decision on #51576 is reached.
2) merge this for now (after addressing your comments) and refactor / upgrade things in #51576.

What do you all think?

**Comment by shepmaster** [comment]
A drive-by comment from AVR-land...

We think that users would like / need the ability to annotate certain pieces of memory with which address space they should go in. We opened [an issue](https://github.com/avr-rust/rust/issues/84) about this, but in short, something like this strawman syntax could address the issue:

```rust
#[space(progmem)]
pub static FOOBAR: &'static str = "hello world";
```

I don't know that this PR would handle something like that, but it may be worth keeping in mind for the underlying design.

**Comment by TimNN** [comment]
I've updated this PR to address the review comments. However I'll close this for now until there's a decision on ##51576.

**Comment by eddyb** [comment]
> I'm not sure what you think, but I feel like @DiamondLovesYou's PR #51576 is the way to go in the long term.

This is a much smaller change, and I'd rather include this now, before we can figure out #51576.

**Review by eddyb — APPROVED** [review]

**Comment by eddyb** [comment]
r=me with rebase

**Comment by pnkfelix** [comment]
Visited for T-compiler triage. This no longer appears to be blocked on waiting for our team.

**Comment by kennytm** [comment]
@bors retry 

3 hour timeout in `x86_64-msvc`, likely due to LLVM cache purge.

**Comment by TimNN** [comment]
@bors retry

* Another timeout on AppVeyor due to LLVM rebuilding?

**Pull request merged**
